### PR TITLE
Fix merkle opening circuit

### LINKED ISSUE #181: Fix merkle opening circuit
**Describe what you want implemented** 
Fix the merkle opening circuit and add tests that catch this in the future

**Describe "Why" this is needed**
The circuit would be valid for an invalid path

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by moCello**
Fix merkle opening circuit and add tests to catch this bug in the future.

Resolves: #181

### CODE REVIEW

File: src/tree/zk.rs [hunk 1]
```diff
@@ -15,65 +15,237 @@ pub fn merkle_opening<const DEPTH: usize>(
...
-    // permutate the level and set the needle to the next level
-    // to the poseidon result of the permutation
-    branch.as_ref().iter().for_each(|level| {
         // Create the bits representation of the offset as witness
+        // and make sure that offset points to a hash in the level
         let offset_flag = level.offset_flag();
-
-        let mut sum = zero;
-        let mut offset_bits = [zero; dusk_hades::WIDTH - 1];
+        let mut sum = TurboComposer::constant_zero();
```
Comment (vlopes11) [inline src/tree/zk.rs:28]: The original purpose of this block was to validate that we have only zeroes/ones, and that we have one, and only one `1`; the rest being `0`.

It was written before we had [component_boolean](https://docs.rs/dusk-plonk/0.12.0/dusk_plonk/constraint_system/struct.TurboComposer.html#method.component_boolean) available, which now completely supersede this implementation. What we need is to run the boolean component for every of these bits (so we guarantee they are either `0` or `1`, and check that they add to `1` (so we guarantee that we have one, and only one `1`).

File: src/tree/zk.rs [hunk 2]
```diff
@@ -15,65 +15,237 @@ pub fn merkle_opening<const DEPTH: usize>(
...
-        level
-            .as_ref()
-            .iter()
-            .zip(perm.iter_mut())
-            .enumerate()
-            .for_each(|(i, (l, p))| {
-                *p = composer.append_witness(*l);
+        // Calculate the root for the next level
+        GadgetStrategy::gadget(composer, &mut container);
+        root = container[1];
```
Comment (vlopes11) [inline src/tree/zk.rs]: Doesn't this look like a [fold](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold) operation? Meaning: you initialize the `root` with `leaf`, mutate and return it for every iteration, and finally return the resulting `root`

File: src/tree/branch.rs [hunk 3]
```diff
@@ -18,24 +18,25 @@ use nstack::NStack;
 /// Represents a level of a branch on a given depth
 #[derive(Debug, Default, Clone, Copy, Canon)]
 pub struct PoseidonLevel {
-    level: [BlsScalar; dusk_hades::WIDTH],
+    pub(crate) level: [BlsScalar; dusk_hades::WIDTH],
```
Comment (ureeves) [inline src/tree/branch.rs:36]: Since the tests have been pulled to unit level, does this need to be `pub(crate)`?

**Review by ureeves — APPROVED** [review]
LGTM. One nit

File: src/tree/zk.rs (hunk 1 above)
Comment (moCello) [reply src/tree/zk.rs:28]: Good point! I'll address this in a new issue #188

File: src/tree/zk.rs (hunk 2 above)
Comment (moCello) [reply src/tree/zk.rs]: done, not sure if it generally becomes more readable with a `fold` but that is just my personal preference.

File: src/tree/branch.rs (hunk 3 above)
Comment (moCello) [reply src/tree/branch.rs:36]: Yes, otherwise the tests can not alter the levels to create a faulty branch.

File: src/tree/zk.rs [hunk 4]
```diff
@@ -15,65 +15,227 @@ pub fn merkle_opening<const DEPTH: usize>(
...
+                // at the level offset
+                let constraint =
+                    Constraint::new().mult(1).a(bit).b(container[i]);
+                let expected = composer.gate_mul(constraint);
+                let constraint = Constraint::new().mult(1).a(bit).b(root);
+                let calculated = composer.gate_mul(constraint);
+
+                // Make sure that the hash at the offset of the current level
+                // matches the hash calculated in the previous level
+                composer.assert_equal(expected, calculated);
```
Comment (vlopes11) [inline src/tree/zk.rs:61]: The `expected` here is a new witness, which can be anything. We need a way to assert that this new witness corresponds to the previous root

**Review by vlopes11 — CHANGES_REQUESTED** [review]
One last question

**Review by vlopes11 — APPROVED** [review]
Never mind the previous review; I overlooked one of the constraints. lgtm!

**Pull request merged**